### PR TITLE
refactor(backend-common/GithubUrlReader): only call fetch inside fetchResponse

### PR DESCRIPTION
**Pull request by secustor**
## Hey, I just made a Pull Request!

This PR centralizes calling `fetch` inside `GithubUrlReader`. 
The purpose is to allow easier customising ( e.g. response caching ) by overriding `fetchResponse` and can be seen as preparation for other refactoring. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by secustor** [comment]
https://github.com/backstage/backstage/pull/21800#issuecomment-1850606065 suggests that the failing pipeline is not triggered by this change.

**Comment by benjdlambert** [comment]
@secustor sorry, I haven't forgotten about this - just been a busy few weeks with the runup to christmas! Will try and get around to this in the next few days!

File: packages/backend-common/src/reading/GithubUrlReader.ts
Comment (freben) [inline packages/backend-common/src/reading/GithubUrlReader.ts:315]: Why was this changed? It changes the public interface, which we probably don't want

Comment (secustor) [reply packages/backend-common/src/reading/GithubUrlReader.ts:315]: To allow overwriting this via an class extension. 

Currently we create our own GithubUrlReader which overwrites `readUrl` to implement caching. 
With this change this low level function can implement caching and reuse the logic in the higher level functions like `readUrl` or `readTree`

**Comment by secustor** [comment]
@benjdlambert friendly ping as this PR is now stale again.

File: packages/backend-common/src/reading/GithubUrlReader.ts
Comment (freben) [reply packages/backend-common/src/reading/GithubUrlReader.ts:315]: I'd prefer to not do that. We pretty strongly favor composition over inheritance, and this is not a significant savings it seems to me. @backstage/maintainers

Comment (secustor) [reply packages/backend-common/src/reading/GithubUrlReader.ts:315]: In that case, what are your thougths in creating a http interface in `backend-common`, which will be used instead of fetch. 

This interface can be than overwritten by devs or customized to implement low level caching, observability and such.

Comment (benjdlambert) [reply packages/backend-common/src/reading/GithubUrlReader.ts:315]: We do have `fetchApiRef` in the frontend, maybe this does make some sense in the backend eventually, but it's a gonna be pretty hard I think to motivate plugin authors and prove the value of it if it's just a wrapper around `fetch`.

@freben what's your thoughts here? Don't have a strong opinion either way.

**Comment by benjdlambert** [comment]
@secustor sorry, I've been ill past few days, just slowing catching back up again.

File: packages/backend-common/src/reading/GithubUrlReader.ts
Comment (secustor) [reply packages/backend-common/src/reading/GithubUrlReader.ts]: I have removed the API change, so this should unblock this PR. 

Currently we are using our own custom GitHubUrlReader, but we would prefer to use an upstream solution and to invest the engineering time rather there.

**Review by freben — APPROVED** [review]
Thanks!

**Review by benjdlambert — APPROVED** [review]
Nice! Let's go with this! :tada: Thanks!

**Pull request merged**
